File: pkg/tproto/checksum.go
package tproto

import (
	"fmt"
	"github.com/zeromicro/go-zero/core/logx"
	"google.golang.org/protobuf/reflect/protoreflect"
	"google.golang.org/protobuf/reflect/protoregistry"
	"hash/crc32"
	"math"
)

var tProtoCrc map[protoreflect.FullName]TConstructor
var CrcTProto map[TConstructor]protoreflect.Message

func InitChecksum() {
	tProtoCrc = make(map[protoreflect.FullName]TConstructor)
	CrcTProto = make(map[TConstructor]protoreflect.Message)

	protoregistry.GlobalTypes.RangeMessages(func(mt protoreflect.MessageType) bool {
		name := mt.Descriptor().FullName()
		checksum := crc32.ChecksumIEEE([]byte(name))
		var crc32Int TConstructor

		if checksum > math.MaxInt32 {
			crc32Int = TConstructor(checksum - math.MaxUint32 - 1)
		} else {
			crc32Int = TConstructor(checksum)
		}
		CrcTProto[crc32Int] = mt.New()
		tProtoCrc[name] = crc32Int
		logx.Debugf(fmt.Sprintf("crc[%d] is mapping to name [%s]", crc32Int, mt.Descriptor().FullName()))
		return true
	})
	logx.Infof("[tProtoInit] Loaded types: %d", len(tProtoCrc))
}

func GetTProtoChecksum(obj TObject) TConstructor {
	if checksum, find := tProtoCrc[obj.ProtoReflect().Descriptor().FullName()]; find {
		return checksum
	}
	return -1
}

func NewTObject(checksum TConstructor) TObject {
	if message, find := CrcTProto[checksum]; find {
		return message.Interface()
	}
	return nil
}

File: app/web/components/TProto/checksum.ts
import {MessageType} from "@protobuf-ts/runtime";
import * as crc32 from "crc-32";
import * as allTypes from "./pb/gen";
import {TConstructor, TObject} from "@/components/TProto/manually_types";

const tProtoCrc = new Map<string, TConstructor>();
const crcTProto = new Map<TConstructor, MessageType<any>>();

export function initChecksum() {
  tProtoCrc.clear();
  crcTProto.clear();

  for (const type of Object.values(allTypes)) {
    const fullName = type.typeName;
    const checksum = crc32.str(fullName) >>> 0;
    const crc32Int: TConstructor = checksum > 0x7FFFFFFF
      ? checksum - 0x100000000
      : checksum;

    tProtoCrc.set(fullName, crc32Int);
    crcTProto.set(crc32Int, type);

    console.info(`crc[${crc32Int}] → ${fullName}`);
  }

  console.info(`[tProtoInit] Loaded types: ${tProtoCrc.size}`);
}

export function getTProtoChecksum(obj: TObject): TConstructor {
  const checksum = tProtoCrc.get(obj.typeName);
  return checksum !== undefined ? checksum : -1;
}

export function newTObject(checksum: TConstructor): MessageType<any> | null {
  return crcTProto.get(checksum) ?? null;
}

export function getTypeNameByChecksum(checksum: TConstructor): string | null {
  const type = crcTProto.get(checksum);
  return type?.typeName ?? null;
}
